### PR TITLE
Scix 170 author network graph using d 3

### PR DESCRIPTION
**Pull request by shinyichen**
depends on #231 

# Updated D3 graph for author network
* Replaced sunburst graph with `visualizations/Graphs/NetworkGraph`
* `visualizations/Graphs/useNetworkGraph` creates the d3 functions to calculate the graph elements for rendering

Screen Shot 2022-07-27 at 6 36 32 PM
Screen Shot 2022-07-27 at 6 37 16 PM

### CODE REVIEW

**Review by thostetler — CHANGES_REQUESTED** [review]
Good job on this, I know it was a ton of work 👍 

I have a few things I think we should work on. 

- For the main container, we can pull out the logic into a reducer to help separate that state management, since really it's the top-level component that passes props around.

- In a few areas there are some mutating logic that I think in general we should try to avoid since it can introduce bugs, but probably fine for these helper functions, since the values are ephemeral.

-  Playing with the author network itself, I noticed that there is a long lag after changing between views (author occurences, paper citations, paper downloads).  Not just a lag though, it's blocking on the main thread, so this one needs to be fixed.

- There is quite a bit of prop drilling that is happening.  I think you could move some of this into it's own store slice in the global store.  I'm not sure how much unnecessary rendering is happening though, just an observation!

Overall I think this looks pretty good and works well! I would be happy to help with any changes you decide you want to make.  

We can also discuss merging and then doing a refactor since there are a bunch of other things added/changed in here beyond the author network.

File: src/components/Expandable/Expandable.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { Box, Collapse, Flex, Stack, StackProps, Text } from '@chakra-ui/react';
+import { ChevronDownIcon, ChevronRightIcon } from '@chakra-ui/icons';
+
+export interface IExpandableProps extends StackProps {
+  title: string;
+  description: string | ReactElement;
+}
+
+export const Expandable = ({ title, description, ...stackProps }: IExpandableProps): ReactElement => {
+  const [isOpen, setIsOpen] = useState(false);
```
[the remark this answers is not in the document]
Comment (shinyichen) [reply src/components/Expandable/Expandable.tsx]: Updated

File: src/components/Tags/Tags.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  return (
+    <Flex justifyContent="space-between" borderWidth={1} p={2} minH={12} {...boxProps}>
+      <Box>
+        {tagItems.length === 0 && placeHolder ? <Text color="gray.300">{placeHolder}</Text> : null}
+        {tagItems.map((item) => (
+          <Tag key={`tag-${item.id}`} size="md" m={1}>
+            <TagLabel>{item.label}</TagLabel>
+            <TagCloseButton onClick={() => handleRemoveItem(item)} />
```
[the remark this answers is not in the document]
Comment (shinyichen) [reply src/components/Tags/Tags.tsx]: Updated

File: src/components/Visualizations/Containers/AuthorNetworkPageContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    { enabled: !!query && !!query.q && query.q.length > 0 },
+  );
+
+  // tranform query data to a list of bibcodes
+  const bibcodes = useMemo(() => {
+    if (bibsQueryResponse) {
+      return bibsQueryResponse.docs.map((d) => d.bibcode);
+    } else {
+      return [];
+    }
```
[the remark this answers is not in the document]
Comment (shinyichen) [reply src/components/Visualizations/Containers/AuthorNetworkPageContainer.tsx]: Updated

File: src/components/Visualizations/GraphPanes/NetworkGraphPane.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import { Radio, RadioGroup, Stack, Text, Input, Button, FormControl, FormLabel, Switch } from '@chakra-ui/react';
+import { ChangeEvent, KeyboardEvent, ReactElement, useState } from 'react';
+import { NetworkGraph } from '../Graphs/NetworkGraph';
+
+export interface INetworkGraphPaneProps {
+  root: IADSApiVisNode;
+  link_data: number[][];
+  views: IView[];
+  onClickNode?: (node: IADSApiVisNode) => void;
+  onChagePaperLimit: (limit: number) => void;
```
[the remark this answers is not in the document]
Comment (shinyichen) [reply src/components/Visualizations/GraphPanes/NetworkGraphPane.tsx]: Fixed

File: src/components/Visualizations/Graphs/NetworkGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+      .style('font-size', (d) => fontSize(d, key))
+      .attr('transform', labelTransform)
+      .attr('text-anchor', textAnchor)
+      .transition()
+      .duration(1500)
+      .attr('opacity', 1);
+  }, []);
+
+  // transition links for view change
+  const transitionLinks = useCallback((links: ILink[]) => {
```
[the remark this answers is not in the document]
Comment (shinyichen) [reply src/components/Visualizations/Graphs/NetworkGraph.tsx]: Removed

**Comment by shinyichen** [comment]
@thostetler I'd like to get your opinion on this topic:
> * There is quite a bit of prop drilling that is happening.  I think you could move some of this into it's own store slice in the global store.  I'm not sure how much unnecessary rendering is happening though, just an observation!

Looking at my code, here are where the prop drillings happen:
1) Main container -> network graph pane -> network graph
    The data passed down are the parts of the response used for the graph and limit. I definitely don't want the base component (network graph) to have any knowledge of where this data comes from, it should just use whatever parent passes so it is reusable. 
2) Main container -> network details
    Passing the summary graph and selected node

Based on that, I don't feel too much of prop drilling here.  Do you still recommend moving some of the data into store?

**Comment by thostetler** [comment]
Yeah you're right it's not a lot.  I think my main concern is that any changes to that root state is going to re-render several large components possibly unnecessarily.
If you don't notice any real extra re-rendering then we can leave it for now. 
I would recommend though rather than passing down lots of handlers/state.  Make a reducer and pass the piece of state you need and the dispatch function to the children.  Then you can do whatever updates without the need for the extra handlers

**Comment by shinyichen** [comment]
I've updated the code with the reducer (will be in a different PR). And I wonder if the main thread blocking issue was from my old graph code, not the D3 code. I did have problem with the old graph switching view, but with D3 there should be no lag.

**Comment by shinyichen** [comment]
https://user-images.githubusercontent.com/636361/183499008-b8ce24f3-fa5e-43d3-b9a6-002faeddafa2.mov

**Pull request merged**